### PR TITLE
Self caching

### PR DESCRIPTION
**Pull request by Sentient07**
This PR is about caching of different Ops, done from the Op class.

### CODE REVIEW

File: theano/gof/op.py [hunk 1]
```diff
@@ -779,10 +780,33 @@ class Op(utils.object2, PureOp, CLinkerOp):
...
+            [key.append(v) for k,v in kwargs.iteritems()]
+
+        # not the best solution, a hack for now
+        for k in key:
+            if isinstance(k, collections.Hashable):
+               key.remove(k) 
+
+        # Since list is not hashable
+        key = tuple(key)
+        if key not in Op.instances:
```
Comment (Sentient07) [inline theano/gof/op.py]: All the objects in the key are not Hashable. What to do in such cases ? Do i ignore and generate a new instance of the class and return ?

Comment (Sentient07) [inline theano/gof/op.py]: Small edit, this comment is for line 797

File: theano/gof/op.py [hunk 2]
```diff
@@ -779,10 +780,33 @@ class Op(utils.object2, PureOp, CLinkerOp):
...
+            Op.instances[key] = object.__new__(cls)
+            if args:
+                for a in args:
+                    Op.instances[key].a = a
+            if kwargs:
+                for k,v in kwargs.iteritems():
+                    Op.instances[key].v = v
```
Comment (Sentient07) [inline theano/gof/op.py]: this should be `Op.instances[key].v = v` . I made this change locally

File: theano/gof/op.py (hunk 1 above)
Comment (nouiz) [inline theano/gof/op.py]: Yes, but print a warning. We want keys and values of props to be hashable. There was 1 excetion I think, but I forgot which one. Can you tell me which one don't hash?

File: theano/gof/op.py [hunk 3]
```diff
@@ -779,10 +780,33 @@ class Op(utils.object2, PureOp, CLinkerOp):
...
-        obj = object.__new__(cls)
+        key = list()
+        if args:
+            [key.append(i) for i in args]
+        if kwargs:
+            [key.append(v) for k,v in kwargs.iteritems()]
+
+        # not the best solution, a hack for now
+        for k in key:
+            if isinstance(k, collections.Hashable):
```
Comment (Sentient07) [inline theano/gof/op.py]: Another silly error.  should have been `not isinstance`. It is like that in my local file though.

**Comment by nouiz** [comment]
killing travis as you will make follow up commit in this PR.

File: theano/gof/op.py [hunk 4]
```diff
@@ -779,10 +780,33 @@ class Op(utils.object2, PureOp, CLinkerOp):
...
+
     def __new__(cls, *args, **kwargs):
         # this function exists to silently and transparently ensure that all
         # existing Ops get a _op_use_c_code attribute
-        obj = object.__new__(cls)
+        key = list()
+        if args:
+            [key.append(i) for i in args]
+        if kwargs:
+            [key.append(v) for k,v in kwargs.iteritems()]
```
Comment (nouiz) [inline theano/gof/op.py]: The key, value is the props, not what it received as paremeter to \__new__

Comment (nouiz) [inline theano/gof/op.py]: In fact, I forgot what args and kwargs would be here. Are they what will be passed to \__init__? If so, then we can't use them. We need to get the \__props__. Can you make a break point here and check if you can access the props?

File: theano/gof/op.py (hunk 1 above)
Comment (Sentient07) [inline theano/gof/op.py]: It is the `Elemwise` Op that has seem to have a dict param (`{0: 0}`). I guess it is the `scalar_op` param ?

Comment (Sentient07) [inline theano/gof/op.py]: Yeah, i remember reporting this previously when i was trying to hash GpuElemwise. You recollect ? And I think that param could be anything (Any Op from `theano.scalar.basic`) and not only a dict as i reported above/

File: theano/gof/op.py (hunk 4 above)
Comment (Sentient07) [inline theano/gof/op.py]: All the param passed to the __init__ will be recieved here as well. So, `(*args, **kwargs)` is a more generalized way. 

> We need to get the `__props__`.

Meaning, iterating through cls.__props__ ?

**Comment by nouiz** [comment]
I talked with @abergeron about this. We don't think it can be done by just implementing Op.__new__. We see 2 options:

1) using mataclass. Arnaud could do it in a short time, but this is pretty advanced stuff.
2) Using a decorator on a python function like:
class MyOp():
   ...
@cache_instance
def my_op(...):
   return MyOp(...)

2) isn't complicated, but would ask that we go over all Op in Theano (and possibly outside). This isn't great and would take much longer then 1). So I think for now we should way to see if @abergeron can implement 1) next week.

**Comment by nouiz** [comment]
killing travis as this isn't the way to go.

**Comment by Sentient07** [comment]
@abergeron If I understand correctly, we'd like to make the `class Op` a singleton.  I've a basic idea of how to go about this. If i'm correct with few correctable mistakes, I'd like to proceed by myself on this. 
```
class ParentOp(type):
    _instances = {}
    # This is done in generalized way of *args and **kwargs
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            # Later, this check could be done using cls._props_dict() 
            cls._instances[cls] = super(ParentOp, cls).__call__(*args, **kwargs)
        return cls._instances[cls]
 
# The existing Op class in theano
class Op():
    __metaclass__ = ParentOp
```
Before i go any further, i'm confused as to whether it should be `PureOp` or `Op` class ? (I guess it should be the Op that should be having the metaclass.
Also, regarding the complexity of handling parameters as @nouiz said, i think after implementing `__props__`  consistently, that concern wouldn't be there either I guess ?

**Comment by nouiz** [comment]
this work is continued in https://github.com/Theano/Theano/pull/5579, so closing.

**Pull request closed without merging**